lib/cache/cache.go

```diff
@@ -154,7 +159,12 @@ func (c *Cache) GetMaybe(key string) (value interface{}, found bool) {
 // Returns true if the entry was found
 func (c *Cache) Delete(key string) bool {
 	c.mu.Lock()
-	_, found := c.cache[key]
+	entry, found := c.cache[key]
+	if found {
+		if err := tryShutdown(entry.value); err != nil {
```

Comment:
This then becomes

```go
if c.finalize != nil {
    c.finalize(entry.value)
}
```

I'd probably make sure c.finalize is never nil so points to a no-op function so you can do without the `if c.finalize { }` which will make the code a lot neater.